homeassistant/components/hassio/__init__.py

```diff
@@ -455,4 +476,134 @@ async def async_handle_core_service(call):
+    hass.data[ADDONS_COORDINATOR] = coordinator
+    await coordinator.async_refresh()
+
+    def noop_listener():
+        """Noop listener to force coordinator to refresh."""
+        return
+
+    # add a noop listener to force coordinator to refresh since entities are disabled by default
+    hass.data[ADDONS_COORDINATOR_UNSUB_LISTENER] = coordinator.async_add_listener(
+        noop_listener
+    )
+
+    for platform in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(config_entry, platform)
+        )
+
+    return True
+
+
+async def async_unload_entry(
+    hass: HomeAssistantType, config_entry: ConfigEntry
+) -> bool:
+    """Unload a config entry."""
+    # Unsubscribe from coordinator listener
+    if ADDONS_COORDINATOR_UNSUB_LISTENER in hass.data:
+        hass.data[ADDONS_COORDINATOR_UNSUB_LISTENER]()
+        hass.data.pop(ADDONS_COORDINATOR_UNSUB_LISTENER)
+
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(config_entry, platform)
+                for platform in PLATFORMS
+            ]
+        )
+    )
+
+    # Pop add-on data
+    hass.data.pop(ADDONS_COORDINATOR, None)
+
+    return unload_ok
+
+
+@callback
+def async_register_addons_in_dev_reg(
+    entry_id: str, dev_reg: DeviceRegistry, addons: List[Dict[str, Any]]
+) -> None:
+    """Register addons in the device registry."""
+    for addon in addons:
+        dev_reg.async_get_or_create(
+            config_entry_id=entry_id,
+            identifiers={(DOMAIN, addon[ATTR_SLUG])},
+            manufacturer=addon.get(ATTR_REPOSITORY) or addon.get(ATTR_URL) or "unknown",
+            model="Home Assistant Add-on",
+            sw_version=addon[ATTR_VERSION],
+            name=addon[ATTR_NAME],
+            entry_type=ATTR_SERVICE,
+        )
+
+
+@callback
+def async_remove_addons_from_dev_reg(
+    dev_reg: DeviceRegistry, addons: List[Dict[str, Any]]
+) -> None:
+    """Remove addons from the device registry."""
+    for addon_slug in addons:
+        if dev := dev_reg.async_get_device({(DOMAIN, addon_slug)}):
+            dev_reg.async_remove_device(dev.id)
+
+
+class HassioAddonsDataUpdateCoordinator(DataUpdateCoordinator):
+    """Class to retrieve Hass.io addons status."""
+
+    def __init__(
+        self, hass: HomeAssistant, config_entry: ConfigEntry, dev_reg: DeviceRegistry
+    ) -> None:
+        """Initialize coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=HASSIO_UPDATE_INTERVAL,
+            update_method=self._async_update_data,
+        )
+        self.data = {}
+        self.entry_id = config_entry.entry_id
+        self.dev_reg = dev_reg
+
+    async def _async_update_data(self) -> Dict[str, Any]:
+        """Update data via library."""
+        data = await self.hass.data[DOMAIN].get_addons_info()
```

Comment:
This returns _all_ add-ons, we only want the installed add-ons right? `/supervisor/info` returns a dictionary with an "addons" key, that contains a list of installed addons.
<https://developers.home-assistant.io/docs/api/supervisor/endpoints#supervisor>

Reply:
ohh good to know, thanks!